File: Online Judges/Huxley/971UmProblemaDeLogistica.py
from heapq import *
import sys
sys.setrecursionlimit(2**20)
DEBUG = 0
inf = 2**33

def prim(graph, cost, visited, start):
    total = 0
    cost[start] = 0
    pq = []
    heappush(pq, [cost[start], start])
    while (pq):
        c, v = heappop(pq)
        if (visited[v]): continue
        visited[v] = 1
        total += c
        for u in graph[v]:
            cost[u[0]] = u[1]
            heappush(pq, [cost[u[0]], u[0]])
    return(total)

tests, run = int(input()), 0
while (run < tests):
    if (run > 0):
        trash = input()
    numWarehouses = int(input())
    intToPoint, pointToInt = {}, {}
    for i in range(numWarehouses):
        point = list(map(int, input().split()))
        pointToInt[(point[0], point[1])] = i
        intToPoint[i] = point
    graph = [[] for i in range(numWarehouses)]
    for i in range(numWarehouses):
        for j in range(numWarehouses):
            dist = abs(intToPoint[i][0] - intToPoint[j][0]) + abs(intToPoint[i][1] - intToPoint[j][1])
            graph[i] += [[j, dist]]
    if (DEBUG):
        print("graph", graph)

    visited, cost = [0] * numWarehouses, [inf] * numWarehouses
    minCost = prim(graph, cost, visited, 0)

    print(run, ": ", minCost, sep='')
    run += 1

File: Online Judges/Huxley/971UmProblemaDeLogistica.cpp
#include <bits/stdc++.h>
using namespace std;
int inf = INT_MAX;

int prim(vector<pair<int, int> > graph[], int cost[], int visited[], int start)
{
  int total = 0;
  cost[start] = 0;
  priority_queue<pair<int, int>, vector<pair<int, int> >, greater<pair<int, int> > > pq;
  pq.push({cost[start], start});
  while (!pq.empty())
  {
    int c = pq.top().first, v = pq.top().second; pq.pop();
    if (visited[v]) continue;
    visited[v] = 1;
    total += c;
    for (auto u : graph[v])
    {
      cost[u.first] = u.second;
      pq.push({cost[u.first], u.first});
    }
  }
  return(total);
}

int main()
{
  int tests, run = 0; scanf("%d", &tests);
  while (run < tests)
  {
    int numWarehouses; scanf("%d", &numWarehouses);
    //map<pair<int, int>, int> pointToInt;
    //map<int, pair<int, int> > intToPoint;
    int point[numWarehouses][2];
    for (int i = 0; i < numWarehouses; i ++)
    {
      int x, y; scanf("%d %d", &x, &y);
      point[i][0] = x; point[i][1] = y;
      //pointToInt[{x, y}] = i;
      //intToPoint[i] = {x, y};
    }
    vector<pair<int, int> > graph[numWarehouses];
    for (int i = 0; i < numWarehouses; i ++)
      for (int j = 0; j < numWarehouses; j ++)
      {
        //int dist = abs(intToPoint[i].first - intToPoint[j].first) + abs(intToPoint[i].second - intToPoint[j].second);
        int dist = abs(point[i][0] - point[j][0]) + abs(point[i][1] - point[j][1]);
        graph[i].push_back({j, dist});
      }

    int cost[numWarehouses], visited[numWarehouses];
    for (int i = 0; i < numWarehouses; i ++)
    {
      cost[i] = inf; visited[i] = 0;
    }
    int minCost = prim(graph, cost, visited, 0);

    printf("%d: %d\n", run, minCost);
    run ++;
  }
  return(0);
}